### PR TITLE
problem approval functionality added in backend

### PR DESCRIPTION
**Pull request by sean-xcamp**
### List of Changes:

- Added problem approval boolean to model/Problem.java
- Added toggleApproval() setter method in model/Problem.java
- Changed getProblemByDifficulty method to check approval in addition to difficulty (only approved problems can be picked)
- Changed ProblemService, GameSocket, Game, and Problem tests by setting every problem to approved when it's created
- Added an exception that prevents people from approving a problem through editProblem if it has zero test cases (Note: not yet tested if this actually works. It's possible that an empty test case will still make the index of the test case arrayList 1 and bypass this error. Can test in production)

### Notes:

- Tests passed (phew)

### CODE REVIEW

File: src/main/java/com/rocketden/main/service/ProblemService.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/service/ProblemService.java:185]: nit: remove these comments now that the feature has been implemented

File: src/test/java/com/rocketden/main/api/GameTests.java
Comment (alankbi) [inline src/test/java/com/rocketden/main/api/GameTests.java:127]: nit: here and for the other 1-3 createSingleProblem methods, could you add at the bottom something like `assertTrue(response.getApproval)`? Just to verify that this was properly set in the database

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (alankbi) [inline src/test/java/com/rocketden/main/service/ProblemServiceTests.java:179]: nit: remove comment (and the other new comments/commented out lines in the file)

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (alankbi) [inline src/test/java/com/rocketden/main/service/ProblemServiceTests.java:470]: small nit but a suggestion would be to use List as the main type instead of ArrayList

```suggestion
        List<ProblemTestCaseDto> emptyCases = new ArrayList<ProblemTestCaseDto>();
```

File: src/main/java/com/rocketden/main/model/problem/Problem.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/model/problem/Problem.java:76]: nit: I don't think this is used anywhere, so I think we can get rid of it

**Review by alankbi — CHANGES_REQUESTED** [review]
Awesome work on the first PR! I requested changes mostly because of small nits here and there (plus the nit below), but otherwise the PR as a whole looks great 👍 

nit: for 1-2 tests in ProblemTests and ProblemService tests (maybe `createEditDeleteProblemSuccess` and `editProblemSuccess` respectively), could you add a line setting approval to true/false, and an assertTrue/assertFalse at the end ensuring the value was set properly?

I also wanted to suggest adding a test in GameTests for `startGameApprovedProblemNotFound`, but in my PR I'm changing some of that behavior so it might be easier if I handle that test case after this is merged into main


Also, two side notes: 

1. I think you might need to configure your GitHub/local git username or email cause your commits right now are showing up as anonymous
2.  
>Added an exception that prevents people from approving a problem through editProblem if it has zero test cases (Note: not yet tested if this actually works. It's possible that an empty test case will still make the index of the test case arrayList 1 and bypass this error. Can test in production)

One recommendation I have would be to download Postman, which is what we use to test the backend routes (I think a while ago we sent an email invite for our workspace which has all the example requests saved). This helps us test things before production and verify that it all works as expected

File: src/main/java/com/rocketden/main/service/ProblemService.java
Comment (sean-xcamp) [reply src/main/java/com/rocketden/main/service/ProblemService.java]: k

File: src/test/java/com/rocketden/main/api/GameTests.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/api/GameTests.java:127]: k

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/service/ProblemServiceTests.java]: k

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/service/ProblemServiceTests.java]: k

File: src/main/java/com/rocketden/main/model/problem/Problem.java
Comment (sean-xcamp) [reply src/main/java/com/rocketden/main/model/problem/Problem.java]: k

**Comment by zpChris** [comment]
> ### List of Changes:
> * Added problem approval boolean to model/Problem.java
> * Added toggleApproval() setter method in model/Problem.java
> * Changed getProblemByDifficulty method to check approval in addition to difficulty (only approved problems can be picked)
> * Changed ProblemService, GameSocket, Game, and Problem tests by setting every problem to approved when it's created
> * Added an exception that prevents people from approving a problem through editProblem if it has zero test cases (Note: not yet tested if this actually works. It's possible that an empty test case will still make the index of the test case arrayList 1 and bypass this error. Can test in production)
> 
> ### Notes:
> * Tests passed (phew)

Wanted to note that for where you reference file or method names you can place `` around the lines like follows, \`model/Problem.java\`, which makes them code-formatted: `model/Problem.java`.

File: src/main/java/com/rocketden/main/service/ProblemService.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/service/ProblemService.java]: My first though on all the responses lol: [https://www.youtube.com/watch?v=07TinrpKG6o&t=70s](https://www.youtube.com/watch?v=07TinrpKG6o&t=70s).

File: src/main/java/com/rocketden/main/model/problem/Problem.java
Comment (zpChris) [inline src/main/java/com/rocketden/main/model/problem/Problem.java:73]: nit: Remove line.

File: src/main/java/com/rocketden/main/service/ProblemService.java
Comment (zpChris) [inline src/main/java/com/rocketden/main/service/ProblemService.java:110]: very small nit: Add lines around this if statement to separate them from the above for and below if statements.

File: src/main/java/com/rocketden/main/service/ProblemService.java
Comment (zpChris) [inline src/main/java/com/rocketden/main/service/ProblemService.java:105]: Include a check to see if `updatedProblem.getApproval() == null`.

File: src/main/java/com/rocketden/main/service/ProblemService.java
Comment (zpChris) [inline src/main/java/com/rocketden/main/service/ProblemService.java:176]: Thanks for adding this! 👍

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (zpChris) [inline src/test/java/com/rocketden/main/service/ProblemServiceTests.java:279]: Is there a reason you left this comment in? Can you remove it?

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (zpChris) [inline src/test/java/com/rocketden/main/service/ProblemServiceTests.java:333]: I'm not sure why you removed this, so maybe re-examine that - either way, either delete the line or un-comment this out in the code.

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (zpChris) [inline src/test/java/com/rocketden/main/service/ProblemServiceTests.java:347]: Same as above (I'm not sure why you removed this, so maybe re-examine that - either way, either delete the line or un-comment this out in the code).

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (zpChris) [inline src/test/java/com/rocketden/main/service/ProblemServiceTests.java:468]: Just want to note that this shouldn't be necessary, as the test cases are, by default, set to an empty `ArrayList`.

File: src/test/java/com/rocketden/main/util/SocketTestMethods.java
Comment (zpChris) [inline src/test/java/com/rocketden/main/util/SocketTestMethods.java:73]: nit: When you add lines, thoughts on keeping these line spacers between different segments of the code? There are a good number of other places with this as well, if you're open to it you could add those blank lines between the different code sections.

File: src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java
Comment (zpChris) [inline src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java:16]: I just realized that if we include `approval` in the `CreateProblemRequest`, we break one of the invariants that was set in this PR - you added a check in the `editProblem` method that the user cannot set the problem to be approved if there are no test cases. However, through creating the problem, we currently have it set up such that you cannot add any test cases until it's actually created, after which you can. You can see that reflected on our UI in the image below:

Create Problem Page

Since it'd be a long fix, I propose that we just include the `approval` boolean for the `editProblem` endpoint. Sorry @alankbi and I didn't identify this earlier! We could enable the ability to add test cases when creating the problem, but that would take a lot more work and we had reasons for preventing this in the first place (we wanted people to first verify that they had the problem inputs set).

**Review by zpChris — CHANGES_REQUESTED** [review]
Awesome job on the first PR! I left some nits and requested changes, but that's all a part of the process, and won't be the case for every PR (just cause this is the first) 👍 

I identified a potential invariant violation that I outlined in [this comment](https://github.com/rocketden/main/pull/131#discussion_r606728850) - sorry that neither @alankbi nor I saw that initially! This does require updating the PR, but hopefully that's actually a fairly easy task. Let me know if you have any questions!

**Comment by zpChris** [comment]
[Add approved boolean to problems to allow for review process](https://trello.com/c/Ylo486wp/318-add-approved-boolean-to-problems-to-allow-for-review-process)

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/service/ProblemServiceTests.java:333]: I wasn't sure myself, but I reran this test and got some "unncessary stubbing" error at those two lines with the `Mockito.doReturn(...)` function; I vaguely recall working through this with Alan and he said the lines weren't necessary but I'm not 100% certain about that. Either way the test passes without this line so I'll delete it for now
Screen Shot 2021-04-05 at 10 45 24 PM

File: src/test/java/com/rocketden/main/util/SocketTestMethods.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/util/SocketTestMethods.java:73]: Will definitely keep this in mind for future PRs! I looked through some of the files I modified and couldn't find many instances of this, though I'm 100% sure they exist; sorry about that

File: src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java
Comment (sean-xcamp) [reply src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java:16]: Got it, I just delete that `Boolean approval` line I added in `CreateProblemRequest right, or is there more to this fix?

Comment (alankbi) [reply src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java:16]: It might be slightly more challenging than that because in the `createSingleProblem`-type methods (the ones in the test files), it's currently just setting approval to true on creation of the problem, so in order to still have that feature (which allows the tests to pass), you'd need to add an additional edit problem request to set the approval to true.

**Comment by zpChris** [comment]
I went to review this (sorry, after quite the delay!) but I don't see any additional commits - is it possible you forgot to push them to this branch @sean-xcamp?

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (zpChris) [reply src/test/java/com/rocketden/main/service/ProblemServiceTests.java:332]: I think I still see the commented-out Mockito lines - can you remove those?

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (zpChris) [reply src/test/java/com/rocketden/main/service/ProblemServiceTests.java:346]: Same here - just delete that comment?

File: src/test/java/com/rocketden/main/util/SocketTestMethods.java
Comment (zpChris) [reply src/test/java/com/rocketden/main/util/SocketTestMethods.java]: No problem, all good 👍

File: src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/dto/problem/CreateProblemRequest.java:16]: I don't think you've made this change yet @sean-xcamp - the list I think is this:
1. Remove the `approval` field in `CreateProblemRequest.java `.
2. Remove all the relevant approval code for `createProblem` in `ProblemService.java`. (You can look at exactly this link - [https://github.com/rocketden/main/pull/131/files](https://github.com/rocketden/main/pull/131/files) - to simply remove the changes one-by-one.)
3. Remove the approval setting in `SocketTestMethods.java`.

There may be others but that's all I could find, I suggest looking at the same "Files Changed" link [here](https://github.com/rocketden/main/pull/131/files) to revert the relevant changes.

**Review by zpChris — CHANGES_REQUESTED** [review]
Just requesting review so that you can finish the changes from last time - I would usually approve but want to make sure those updates get made. Should be really quick, great job thus far 👍

File: src/test/java/com/rocketden/main/service/ProblemServiceTests.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/service/ProblemServiceTests.java:332]: > I think I still see the commented-out Mockito lines - can you remove those @sean-xcamp? (There are two lines I see 👍 )

Yep, thanks for pointing that out; not sure how I missed it! I included it in the next commit

**Review by zpChris — APPROVED** [review]
LGTM! ✅ 

Full disclosure I didn't re-run all the tests because this looks good, but maybe @alankbi when you end up reviewing it you can run that command? Otherwise np - thanks @sean-xcamp! 👍 

Also, for the PRs following this one, I think likely just @alankbi OR me (@zpChris) could act as reviewers, and maybe @sean-xcamp and @jasonz6688 you can start to review PRs as well, as that will be helpful training for when we may add more people in the future.

**Comment by alankbi** [comment]
Why the tests are most likely failing: when you removed the approval setting from the create problem request, I think it most likely re-introduced the original bug where the helper methods `createSingleProblem` creates an unapproved problem, and none of the `getProblemByDifficultyAndApproval` tests are able to find a problem in the tests. 

So to solve this, I think you'll probably have to edit each of the `createSingleProblem`-type methods to include a second `UpdateSettingsRequest` call that sets the newly created problem to be approved.

**Comment by zpChris** [comment]
> Why the tests are most likely failing: when you removed the approval setting from the create problem request, I think it most likely re-introduced the original bug where the helper methods `createSingleProblem` creates an unapproved problem, and none of the `getProblemByDifficultyAndApproval` tests are able to find a problem in the tests.
> 
> So to solve this, I think you'll probably have to edit each of the `createSingleProblem`-type methods to include a second `UpdateSettingsRequest` call that sets the newly created problem to be approved.

It would be smart to make a helper method that is `createSingleApprovedProblem` so this change is only made once.

Sorry I didn't notice that when approving the PR initially! Thanks for catching that @alankbi

**Comment by sean-xcamp** [comment]
> > Why the tests are most likely failing: when you removed the approval setting from the create problem request, I think it most likely re-introduced the original bug where the helper methods `createSingleProblem` creates an unapproved problem, and none of the `getProblemByDifficultyAndApproval` tests are able to find a problem in the tests.
> > So to solve this, I think you'll probably have to edit each of the `createSingleProblem`-type methods to include a second `UpdateSettingsRequest` call that sets the newly created problem to be approved.
> 
> It would be smart to make a helper method that is `createSingleApprovedProblem` so this change is only made once.

This would result in a lot of duplicate code right @zpChris would that be fine?

**Comment by zpChris** [comment]
> This would result in a lot of duplicate code right @zpChris would that be fine? 

Sorry not sure I follow - wouldn't my solution result in **less** duplicate code, since you're placing the duplicate code about editing the problem to change its approval in a helper method?

**Comment by sean-xcamp** [comment]
> > This would result in a lot of duplicate code right @zpChris would that be fine?
> 
> Sorry not sure I follow - wouldn't my solution result in **less** duplicate code, since you're placing the duplicate code about editing the problem to change its approval in a helper method?

My understanding was that there would be a `createSingleProblem()` method and **another** `createSingleApprovedProblem()` method, with almost all lines of code shared between those two. I'm not exactly sure if I follow if otherwise.

**Comment by zpChris** [comment]
> My understanding was that there would be a `createSingleProblem()` method and **another** `createSingleApprovedProblem()` method, with almost all lines of code shared between those two. I'm not exactly sure if I follow if otherwise.

Ah - what you could do is in the `createSingleApprovedProblem()` method, you could call the  `createSingleProblem()` method to avoid that duplicate code.

**Comment by sean-xcamp** [comment]
> > My understanding was that there would be a `createSingleProblem()` method and **another** `createSingleApprovedProblem()` method, with almost all lines of code shared between those two. I'm not exactly sure if I follow if otherwise.
> 
> Ah - what you could do is in the `createSingleApprovedProblem()` method, you could call the `createSingleProblem()` method to avoid that duplicate code.

Got it, that clarification makes a lot more sense. Will do.

**Comment by zpChris** [comment]
Ok - here's the list of fixes (I found these all through debugging, so that's a _super_ important tool):
- The `RoomSocketTests` were failing because the `createSingleProblemAndTestCases` never actually attached the test cases to the **returned** `problemActual`. Thus, when you tried to update the problem and change the setting to approve, it also registered it as deleting the test cases and setting it to approved. Since you cannot have an approved problem without test cases, this threw an error.
  - You initially tried to solve this issue by adding test cases to the `createSingleApprovedProblemAndTestCases` method; however, the output you set was `"a"`, which was an invalid type since it was supposed to be an Array of Integers, thus it threw an error.
- The `socketReceivesMessageOnStartGame` in `RoomSocketTests` had to change `createSingleProblemAndTestCases` to `createSingleApprovedProblemAndTestCases`.

File: src/main/java/com/rocketden/main/model/problem/Problem.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/model/problem/Problem.java:36]: nit: add a line between these two

File: src/test/java/com/rocketden/main/api/GameTests.java
Comment (alankbi) [inline src/test/java/com/rocketden/main/api/GameTests.java:106]: nit: delete this since it's not being used

File: src/test/java/com/rocketden/main/api/GameTests.java
Comment (alankbi) [inline src/test/java/com/rocketden/main/api/GameTests.java:193]: Just wanted to double-check - do we need to be adding a test case in this method? I thought the `createSingleProblemAndTestCases` method returns a ProblemDto that already has a test case

Same thing for the other `createSingleApprovedProblemAndTestCases` method

**Review by alankbi — APPROVED** [review]
Approved! However I left a few small nits, so be sure to take a look at those before merging :)

File: src/test/java/com/rocketden/main/api/GameTests.java
Comment (sean-xcamp) [reply src/test/java/com/rocketden/main/api/GameTests.java:193]: previously we tried running these tests without this and it failed, so I'm pretty sure it's necessary. It's a little bit odd.

Comment (zpChris) [reply src/test/java/com/rocketden/main/api/GameTests.java:184]: See [this](https://github.com/rocketden/main/pull/131#issuecomment-817464717) as to why that was @sean-xcamp - I don't think it's necessary anymore.

**Pull request merged**
